handbook/engineering/introducing_a_new_service.md

```diff
@@ -0,0 +1,45 @@
+# Introducing a new service
+
+Before reading this document be sure to first check out our [architecture overview](https://docs.sourcegraph.com/dev/background-information/architecture).
+
+## Terminology
+
+When we say "service" here we are referring to code that runs _within a Docker container_. This may be one or more processes, servers, goroutines, threads, etc. operating in a single Docker container.
+
+## When does introducing a new service make sense?
+
+Sourcegraph is composed of several smaller services (gitserver, searcher, symbols, etc.) and a single monolithic service (the frontend).
+
+When thinking of adding a new service, it is important to think through the following questions carefully:
+
+1. Does the code belong in an existing service?
+2. Instead of introducing a new service, could the process/etc reasonably live inside the frontend monolith as a singleton background worker, goroutine, etc.?
```

Comment:
nit: The frontend isn't a singleton itself so we've ended up putting a lot of these singleton background jobs in repo-updater instead.

Reply:
```suggestion
2. Instead of introducing a new service, could the process/etc reasonably live inside another service (the frontend, etc.) as a background worker, goroutine, etc.?
```